### PR TITLE
Introducing Logging for Customer Product Downloads and Advanced Reporting

### PR DESCRIPTION
**Pull request by procifer**
Completes #12517

Adds a new WC_Customer_Download_Log class, and supporting data store, database table, etc.  When customer downloads take place, the events are now tracked in the log table.  Additional reporting in the main section of WC reports is now available to view download logs filtered by a given permission.

Also includes unit tests for the new WC_Customer_Download_Log class, as well as the new functionality in WC_Customer_Download.

This is from my fork of the repo, so is against the master branch.  Once ready for merging, this should likely go into a feature branch.  Also, it targets version 3.3.0, but please review the target version once ready for merging.

### CODE REVIEW

**Comment by claudiosanches** [comment]
This is a little big and there is some changes that does not sounds related to the main issue. So for sure will be hard to review each line of this PR.
Thanks for your help.

**Comment by procifer** [comment]
@claudiosanches thanks for taking a look!  Yea, it's fairly involved so definitely let me know if you have any questions or need me to explain anything in details as you go through.  Also, since I'm on trial, this is my first PR so definitely let me know if there's a better approach to organizing it, i.e. if you think certain parts should be on their own, etc.  Thanks!

File: includes/abstracts/abstract-wc-product.php
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: We don't have phpcs enabled on here yet (holding off until 3.3) but just wanted to point out these types of comments should end with a full stop. &lt;/nitpick&gt; :)

File: includes/abstracts/abstract-wc-product.php
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: Using `empty` negates the need for `isset`.

File: includes/admin/class-wc-admin-reports.php
Comment (mikejolley) [inline includes/admin/class-wc-admin-reports.php]: We don't use title case across our screens.

File: includes/admin/meta-boxes/views/html-order-download-permission.php
Comment (mikejolley) [inline includes/admin/meta-boxes/views/html-order-download-permission.php]: `esc_url` is more appropriate here.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: This appears to be missing localisation and `notice` classes. Maybe use a `wp_die` if you want to stop the rest of the page loading if this occurs?

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: We have a `get_formatted_name` method for products which may work better here.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: No localisation.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: `get_formatted_name`

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: edit_post_link does not need an echo it seems. This error is repeated throughout the PR and needs correction.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: Since this is filterable, let's ensure user cannot set `0` because it's used in division below.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mikejolley) [inline includes/admin/reports/class-wc-report-downloads.php]: Why return empty array? This method does not return anything.

File: includes/class-wc-customer-download-log.php
Comment (mikejolley) [inline includes/class-wc-customer-download-log.php]: Since this is a new class, why does it need ArrayAccess and backwards compatibility? Can you explain what this is replacing in the old codebase?

File: includes/class-wc-customer-download-log.php
Comment (mikejolley) [inline includes/class-wc-customer-download-log.php]: Use Automattic not WooThemes.

File: includes/class-wc-customer-download-log.php
Comment (mikejolley) [inline includes/class-wc-customer-download-log.php]: Weird spacing issue.

File: includes/class-wc-customer-download.php
Comment (mikejolley) [inline includes/class-wc-customer-download.php]: 👍

File: includes/class-wc-customer-download.php
Comment (mikejolley) [inline includes/class-wc-customer-download.php]: I wonder if doing this with PHP may cause issues if downloads are made in quick succession. Would it be more reliable to do this in SQL?

File: includes/class-wc-customer-download.php
Comment (mikejolley) [inline includes/class-wc-customer-download.php]: > I wonder if doing this with PHP may cause issues if downloads are made in quick succession. Would it be more reliable to do this in SQL?

File: includes/class-wc-download-handler.php
Comment (mikejolley) [inline includes/class-wc-download-handler.php]: Indentation looks off.

File: includes/class-wc-download-handler.php
Comment (mikejolley) [inline includes/class-wc-download-handler.php]: Leave space after `!`

File: includes/class-wc-install.php
Comment (mikejolley) [inline includes/class-wc-install.php]: Not a problem afaik but reason?

File: includes/class-wc-install.php
Comment (mikejolley) [inline includes/class-wc-install.php]: We have an issue about out long table names causing issues on some hosts. https://github.com/woocommerce/woocommerce/pull/16390 Maybe this should be kept short to save us needing to renaming in the future. `wc_download_log`? cc @claudiosanches

File: includes/class-wc-post-data.php
Comment (mikejolley) [inline includes/class-wc-post-data.php]: If this method is no longer needed, deprecate it.

File: includes/class-woocommerce.php
Comment (mikejolley) [inline includes/class-woocommerce.php]: Why is `since` being changed? This is rewriting history :p

File: woocommerce.php
Comment (mikejolley) [inline woocommerce.php]: Leave this out please. It will be done before this PR is merged :)

**Review by mikejolley — CHANGES_REQUESTED** [review]
Leaving code review feedback. I've not 'tested' the functionality yet, just gone through the PR :)

File: includes/admin/class-wc-admin-reports.php
Comment (mjangda) [inline includes/admin/class-wc-admin-reports.php]: Minor: should use single quotes here.

File: includes/admin/meta-boxes/views/html-order-download-permission.php
Comment (mjangda) [inline includes/admin/meta-boxes/views/html-order-download-permission.php]: We may want to use `number_format_i18n` here to larger numbers look nice.

File: includes/admin/meta-boxes/views/html-order-download-permission.php
Comment (mjangda) [inline includes/admin/meta-boxes/views/html-order-download-permission.php]: Should we run this value through `rawurlencode` to be on the safe side?

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: Rather than repeating the use of `$_GET['permission_id']` throughout this method, maybe we should assign to a variable instead and use that.

File: tests/unit-tests/customer/customer-download-log.php
Comment (mjangda) [inline tests/unit-tests/customer/customer-download-log.php]: Whenever we have multiple assertions inside one test, it's always a good idea to add the `message` param so that when tests fail, we can more easily pinpoint which one it is (rather than trying to figure out what "expected 0 did not match 1" means).

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: This is a really long `if`; would be good to break it up. (Generally, it's preferred not to do too much computation within `if` clauses anyway; better to assign to a var and check the var.)

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: Should we have a convenience method for getting the table name and maybe move `woocommerce_downloadable_product_download_log` to a const?

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: Do we need to handle this if/when it fails?

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: Should this action name be a bit more specific  (i.e. indicate that we're inserting)?

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: Should these action names be a bit more specific (i.e. indicate that we're inserting)?

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: Should we add a type hint for this param? (And do the same for the other methods in this class)

File: includes/data-stores/class-wc-customer-download-data-store.php
Comment (mjangda) [inline includes/data-stores/class-wc-customer-download-data-store.php]: `__METHOD__` may be more useful than the hardcoded name.

Should we specify a replacement?

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: To be safe, should wrap `get_edit_user_link` in `esc_url`.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: `esc_url`

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: The building of the title mixed with the HTML makes the code block a bit hard to read/follow and could lead to bugs. Can we separate out the title generation piece (either new function or outside the HTML block)?

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: Move the variable assignment out of the `if`

**Review by mjangda — CHANGES_REQUESTED** [review]

File: includes/admin/class-wc-admin-reports.php
Comment (procifer) [inline includes/admin/class-wc-admin-reports.php]: Ah I see; I'll switch that.  I was basing this on the "Most Stocked" one so I followed the same case.  I'll change that to "Most stocked" while I'm in there.

**Comment by procifer** [comment]
@mikejolley and @mjangda, thanks for the code reviews; great feedback!  I will be working on these changes this week and hope to have them wrapped up soon.  Thanks!

File: includes/admin/reports/class-wc-report-downloads.php
Comment (procifer) [inline includes/admin/reports/class-wc-report-downloads.php]: Ah, good catch - all the details at the top of the report need to be localized.  I'll try changing this info to model after the top of the order details which works a little better than this sentence format.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (procifer) [inline includes/admin/reports/class-wc-report-downloads.php]: Actually I think the edit_post_link function emits the full `A` tag with the url included and passed through `esc_url` already, so should be good on this one (let me know if I'm misunderstanding though).  I'll remove the `echo` though since it's already included.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (procifer) [inline includes/admin/reports/class-wc-report-downloads.php]: Good point!  I'll move that out to a separate function to keep it organized.

File: includes/admin/reports/class-wc-report-downloads.php
Comment (procifer) [inline includes/admin/reports/class-wc-report-downloads.php]: Ah, you're right; will change it to just leave the member variable `items` empty and a simple return.

File: includes/class-wc-customer-download-log.php
Comment (procifer) [inline includes/class-wc-customer-download-log.php]: @mikejolley good question.  It seemed that all `WC_Data` classes had this interface implemented (e.g. `WC_Customer_Download` which I was basing this on), so I assumed there was a spot in the codebase that required this.  I'll try removing this and will make sure tests and such still pass.

File: includes/class-wc-install.php
Comment (procifer) [inline includes/class-wc-install.php]: Ah, interesting.  I'll switch it to `wc_download_log` for now.  Let me know if you guys think we should adjust.  Thanks!

File: includes/class-wc-install.php
Comment (procifer) [inline includes/class-wc-install.php]: Yea, this is due to the longer UUIDs we'll be using, which are 36 chars long, instead of the 32 char MD5 previously.

File: includes/class-woocommerce.php
Comment (procifer) [inline includes/class-woocommerce.php]: Oops, that's definitely a mistake; reverting that back 😄

File: includes/data-stores/class-wc-customer-download-data-store.php
Comment (procifer) [inline includes/data-stores/class-wc-customer-download-data-store.php]: Ah, good point; I'll switch that.  On the replacement, I don't think there's really a replacement in this case since we shouldn't really be changing IDs at all anymore in the new approach.  So instead of replacing calls to this, they simply shouldn't be trying to change IDs.

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (procifer) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: Yea, that's a great idea, especially given @mikejolley's  comments above about possibly shortening table names for compatibility.  I'll make this a constant and will replace all hardcoded references to it.

File: includes/data-stores/class-wc-customer-download-log-data-store.php
Comment (procifer) [inline includes/data-stores/class-wc-customer-download-log-data-store.php]: That's a good point - I'll add a `wp_die` with relevant messaging just in case something fails during the insert.  That should stop the download endpoint to make sure downloads don't continue until they're being tracked properly.

**Comment by procifer** [comment]
@mikejolley and @mjangda, thanks again for the reviews!  I submitted some new changes which should address all the code review feedback.  All unit tests should be passing and I tested new installations as well as upgrades, so things should be working properly.  Let me know what next steps should be and/or if you see any other revisions to work on.  Thanks!

**Review by mikejolley — APPROVED** [review]

**Comment by mikejolley** [comment]
Code looks good to me. I've requested a final review from @claudiulodro to check over everything.

Few minor changes after seeing how this all works:

- In the download report, when there is no user, output "guest".
- Would it be useful to break out some stats on the main download reports page e.g. most downloaded, most active IP/user, with the ability to view by those too?
- I think we should tweak the wording and tab on which this report is found. I'm wondering if `products` is really the most appropriate place? Suggestions?
- I wonder if we should call them 'customer downloads' to reflect that these are downloads from customers and require permissions.

@mjangda

**Comment by procifer** [comment]
@mikejolley Thanks so much for the additional review!  And great ideas.  I can definitely make those revisions.  Since this is approved now, should I make those revisions in a separate PR?  Or would those be appropriate as additional commits on this one?

- Re: breaking out the most downloaded, active IP/user, etc., are you thinking that should appear on the top of the report regardless of the current filter state?  Or what if we have a separate sub-tab for a "Summary" view (similar to how the other report tabs work)?  Then those summary lines could link to a filtered "Detail" view.  This might take a bit more time though; should we put this in a subsequent PR to keep this one a little simpler?  (Happy to put in here too, just wondering about best practice 😄 )

- Excellent question re: the labeling and position in the reports UI.  I struggled with that decision too.  I didn't want to put them under the other taxonomies since it's definitely a separate data set we're talking about, but I totally agree that the "Products" heading isn't perfect either.  I hate to clutter the top level tabs, but maybe just a "Customer Downloads" top level might be most appropriate.  Then we could support those details/summary tabs I was mentioning under that.  Would that work?

Thanks again for the feedback!  Let me know best next steps.

@mjangda

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: Minor: if we expect an integer here should we use `intval` or `absint`?

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: We're using a generic exception here; will that always mean that the permission wasn't found? Or could other errors creep in here too? (Note: I haven't looked too closely into either `WC_Customer_Download` or `wc_get_product`)

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: Minor: for consistency with `order`, can we move this inside the `if` and reverse the condition?

File: includes/admin/reports/class-wc-report-downloads.php
Comment (mjangda) [inline includes/admin/reports/class-wc-report-downloads.php]: To be on the safe side, we should assign `$_GET['permission_id']` to a variable and cast using `intval` or `absint`.

**Review by mjangda — CHANGES_REQUESTED** [review]
Flagged a couple of minor things but overall this looks good.

File: includes/class-wc-customer-download-log.php
Comment (claudiulodro) [inline includes/class-wc-customer-download-log.php]: This method is exactly the same as `WC_Data::save`. You can delete it and just let the parent method do the work.

File: tests/unit-tests/customer/customer-download-log.php
Comment (claudiulodro) [inline tests/unit-tests/customer/customer-download-log.php]: Nitpick: Formatting

**Review by claudiulodro — CHANGES_REQUESTED** [review]
Overall everything looks good and works well. Good job keeping the code consistent with the rest of the WooCommerce code and following WordPress best practices. 

Just a couple small changes that haven't been brought up yet.

**Comment by mikejolley** [comment]
> Re: breaking out the most downloaded, active IP/user, etc., are you thinking that should appear on the top of the report regardless of the current filter state? Or what if we have a separate sub-tab for a "Summary" view (similar to how the other report tabs work)? Then those summary lines could link to a filtered "Detail" view. This might take a bit more time though; should we put this in a subsequent PR to keep this one a little simpler? 

New PR on top of this one may be wise so it can be reviewed separately. I was thinking the 'main' view before filtering could just show show stats like the other overview reporting sections, the difference being rather than charts the main section is a table as you have now? Keeping it to one screen - this is only a small part of WC as a whole.

> Excellent question re: the labeling and position in the reports UI. I struggled with that decision too. I didn't want to put them under the other taxonomies since it's definitely a separate data set we're talking about, but I totally agree that the "Products" heading isn't perfect either. I hate to clutter the top level tabs, but maybe just a "Customer Downloads" top level might be most appropriate. Then we could support those details/summary tabs I was mentioning under that. Would that work?

I'd probably think it fits under either orders or customers. You could argue either way. I think orders.

**Comment by mikejolley** [comment]
Continuing this PR here https://github.com/woocommerce/woocommerce/pull/17679

**Pull request closed without merging**
